internal/client/db/db.go

```diff
@@ -460,10 +462,192 @@ func (bdb *blockchainDB[H, N, E, Header]) Leaves() ([]H, error) {
 	return bdb.leaves.Hashes(), nil
 }
 
-func (bdb *blockchainDB[H, N, E, Header]) DisplacedLeavesAfterFinalizing(blockNumber N) ([]H, error) {
+func (bdb *blockchainDB[H, N, E, Header]) DisplacedLeavesAfterFinalizing(
+	finalizedBlockHash H, finalizedBlockNumber N,
+) (blockchain.DisplacedLeavesAfterFinalization[H, N], error) {
 	bdb.leavesMtx.RLock()
 	defer bdb.leavesMtx.RUnlock()
-	return bdb.leaves.DisplacedByFinalHeight(blockNumber).Leaves(), nil
+	leaves, err := bdb.Leaves()
+	if err != nil {
+		return blockchain.DisplacedLeavesAfterFinalization[H, N]{}, err
+	}
+
+	// If we have only one leaf there are no forks, and we can return early.
```

Comment:
can we add the debug log here ([link](https://github.com/paritytech/polkadot-sdk/blob/fdb4554e26ebdd4d729158501a3ddb3c6ebdfb6f/substrate/primitives/blockchain/src/backend.rs#L266)).